draft-ietf-mls-protocol.md

```diff
@@ -1522,218 +1464,274 @@ the Init message as follows:
+2. A member of the group broadcasts a Commit message that causes one or more
+   proposed changes to enter into effect
 
-In order to add a new member to the group, an existing member of the
-group must take two actions:
+The group thus evolves from one cryptographic state to another each time a
+Commit message is sent and processed.  These states are referred to as "epochs"
+and uniquely identified among states of the group by four-octet epoch values.
+When a new group is initialized, its initial state epoch 0x00000000.  Each time
+a state transition occurs, the epoch number is incremented by one.
 
-1. Send a Welcome message to the new member
-2. Send an Add message to the group (including the new member)
+[[ OPEN ISSUE: It would be better to have non-linear epochs, in order to
+tolerate forks in the history. ]]
 
-The Welcome message contains the information that the new member
-needs to initialize a GroupContext object that can be updated to the
-current state using the Add message.  This information is encrypted
-for the new member using HPKE.  The recipient key pair for the
-HPKE encryption is the one included in the indicated ClientInitKey,
-corresponding to the indicated ciphersuite.  The "add_key_nonce"
-field contains the key and nonce used to encrypt the corresponding
-Add message; if it is not encrypted, then this field MUST be set to
-the null optional value.
+## Proposals
+
+Proposals are included in an MLSPlaintext by way of a Proposal structure that
+indicates their type:
 
 ~~~~~
-struct {
-    HPKEPublicKey public_key;
-    optional<Credential> credential;
-} RatchetNode;
+enum {
+    invalid(0),
+    add(1),
+    update(2),
+    remove(3),
+    (255)
+} ProposalType;
 
 struct {
-    opaque key<0..255>;
-    opaque nonce<0..255>;
-} KeyAndNonce;
+    ProposalType msg_type;
+    select (Proposal.msg_type) {
+        case add:    Add;
+        case update: Update;
+        case remove: Remove;
+    };
+} Proposal;
+~~~~~
 
-struct {
-    ProtocolVersion version;
-    opaque group_id<0..255>;
-    uint32 epoch;
-    optional<RatchetNode> tree<1..2^32-1>;
-    opaque interim_transcript_hash<0..255>;
-    opaque init_secret<0..255>;
-    optional<KeyAndNonce> add_key_nonce;
-} WelcomeInfo;
+### Add
 
+An Add proposal requests that a client with a specified ClientInitKey be added
+to the group.
+
+~~~~~
 struct {
-    opaque client_init_key_id<0..255>;
-    CipherSuite cipher_suite;
-    HPKECiphertext encrypted_welcome_info;
-} Welcome;
+    ClientInitKey init_key;
+} Add;
 ~~~~~
 
-In the description of the tree as a list of nodes, the `credential`
-field for a node MUST be populated if and only if that node is a
-leaf in the tree.
+The proposer of the Add does not control where in the group's ratchet tree the
+new member is added.  Instead, the sender of the Commit message chooses a
+location for each added member and states it in the Commit message.
```

Comment:
That is the desire, but it runs into the problem that "first" isn't well-defined in a context where members might see two Add messages in different orders.  So the idea here is that the Committer acts as the arbiter and decides the order.

Reply:
Doesn't the Commit message define an order over the Adds? You could
```diff
- AddCommit adds<0..2^16-1>;
+ ProposalID adds<0..2^16-1>;
```
because the index is implicit

Reply:
Oh, I see. Good point.  If I try hard, I can imagine scenarios where you would not want that ("Each user gets 8 devices, so we're going to assign each user a slice of the tree"), but for this first cut, I'm happy to have it simpler.  We can add the flexibility back if people ask for it.